docs/architecture.md

```diff
@@ -0,0 +1,438 @@
+# Torch-MLIR Architecture
+
+## Introduction
+
+The Torch-MLIR project provides core infrastructure for bridging between the
+Torch ecosystem and the MLIR ecosystem. Torch-MLIR does not attempt to provide a
+production end-to-end flow for PyTorch programs by itself, but is a useful
+component for constructing one.
+
+## Overview
+
+Torch-MLIR has two parts, which we call the "frontend" and "backend". These two
+halves interface at an abstraction layer which we call the "backend contract"
+and which is a normalized subset of the `torch` dialect.
```

Comment:
If it's not too much work, it might help to accompany this with figures (like the one on the README).